File: companies/eBay/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/description/
#
# algorithms
# Medium (63.63%)
# Likes:    12099
# Dislikes: 360
# Total Accepted:    1.8M
# Total Submissions: 2.8M
# Testcase Example:  '[3,9,20,null,null,15,7]'
#
# Given the root of a binary tree, return the zigzag level order traversal of
# its nodes' values. (i.e., from left to right, then right to left for the next
# level and alternate between).
# 
# 
# Example 1:
# 
# 
# Input: root = [3,9,20,null,null,15,7]
# Output: [[3],[20,9],[15,7]]
# 
# 
# Example 2:
# 
# 
# Input: root = [1]
# Output: [[1]]
# 
# 
# Example 3:
# 
# 
# Input: root = []
# Output: []
# 
# 
# 
# Constraints:
# 
# 
# The number of nodes in the tree is in the range [0, 2000].
# -100 <= Node.val <= 100
# 
# 
#

# @lc code=start
from typing import List, Optional
from collections import deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Interview explanation:
        Use normal BFS to preserve level grouping, then reverse the level values
        on alternating depths. This is simpler and less error-prone than changing
        child enqueue order, because the queue invariant remains standard BFS.

        Edge cases and tests:
        - Empty tree returns [].
        - One level is not reversed.
        - Missing children do not affect level boundaries.

        Complexity: O(n) time, O(w) queue space.
        """
        if not root:
            return []
        ans = []
        q = deque([root])
        left_to_right = True
        while q:
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            ans.append(level if left_to_right else level[::-1])
            left_to_right = not left_to_right
        return ans
# @lc code=end

File: LinkedIn/103.binary-tree-zigzag-level-order-traversal.go
package leetcode

// ZigzagLevelOrder103 performs normal BFS to preserve level grouping, then
// reverses values on alternating levels. The queue invariant stays standard BFS.
//
// Time: O(n). Space: O(w).
func ZigzagLevelOrder103(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	ans := [][]int{}
	q := []*TreeNode{root}
	leftToRight := true
	for len(q) > 0 {
		size := len(q)
		level := make([]int, 0, size)
		for i := 0; i < size; i++ {
			node := q[0]
			q = q[1:]
			level = append(level, node.Val)
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		if !leftToRight {
			for l, r := 0, len(level)-1; l < r; l, r = l+1, r-1 {
				level[l], level[r] = level[r], level[l]
			}
		}
		ans = append(ans, level)
		leftToRight = !leftToRight
	}
	return ans
}
